### PR TITLE
Disallow '!' or '?' at the end of the LHS in an assignment

### PR DESCRIPTION
**Pull request by Maroo-b**
Issue: https://github.com/crystal-lang/crystal/issues/6685
This affects mainly variable defined in the top level, instance and class variable raise already an exception.

I'm curious how do you guys debug the compiler, I'm still new here and I was just adding `puts` statement for debugging, is there any trick to know that could make me or someone else more productive in the future?

### CODE REVIEW

**Comment by asterite** [comment]
We debug with puts like you

File: src/compiler/crystal/syntax/parser.cr [hunk 1]
```diff
@@ -333,6 +332,10 @@ module Crystal
               raise "can't change the value of self", location
             end
 
+            if atomic.is_a?(Call) && atomic.name.ends_with?(/[?!]/)
```
Comment (j8r) [inline src/compiler/crystal/syntax/parser.cr]: Using `ends_with?('?')` and `ends_with?('!')` is faster than a regexp
```cr
require "benchmark"

REG = /[?!]/
STR = "123!"

Benchmark.ips do |x|
  x.report("regexp") do
    STR.ends_with? REG
  end

  x.report("chars") do
    STR.ends_with?('?') || STR.ends_with?('!')
  end
end
```
```
regexp 710.41k (  1.41µs) (± 4.25%)  496 B/op  639.10× slower
 chars 454.02M (   2.2ns) (± 2.09%)    0 B/op         fastest
```

Comment (Maroo-b) [reply src/compiler/crystal/syntax/parser.cr]: @j8r thank you for the benchmarking! 
I'll edit the condition to be  `atomic.name.ends_with?('?') || atomic.name.ends_with?('!')`

**Comment by j8r** [comment]
@Maroo-b there is also the @bcardiff's [puts](https://github.com/crystal-lang/crystal/issues/6615#issuecomment-416740878):  `pp!({x, x.class, typeof(x)})`
One can have:
```cr
def pp_bcardiff(x)
  pp!({x, x.class, typeof(x)}
end
```
Profit :smile:

**Comment by Maroo-b** [comment]
@j8r I was actually searching for this method in the code source 😄 . Maybe we should add for internal use only ⚒

File: src/compiler/crystal/syntax/parser.cr [hunk 2]
```diff
@@ -333,6 +332,11 @@ module Crystal
               raise "can't change the value of self", location
             end
 
+            if atomic.is_a?(Call) && (atomic.name.ends_with?('?') ||
+               atomic.name.ends_with?('!'))
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/parser.cr]: Minor: Logic is easier to follow when both name checks are on the same line:
```suggestion
               (atomic.name.ends_with?('?') || atomic.name.ends_with?('!'))
```

File: src/compiler/crystal/syntax/parser.cr [hunk 3]
```diff
@@ -333,6 +332,11 @@ module Crystal
               raise "can't change the value of self", location
             end
 
+            if atomic.is_a?(Call) && (atomic.name.ends_with?('?') ||
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/parser.cr]: You need to remove the part from this line. Couldn't post a multi-line diff suggestion.

**Comment by Maroo-b** [comment]
Thank you @straight-shoota for the review, I updated the PR as suggested.

File: src/compiler/crystal/syntax/parser.cr [hunk 4]
```diff
@@ -315,7 +315,6 @@ module Crystal
           break
         when :"="
           slash_is_regex!
-
```
Comment (Sija) [inline src/compiler/crystal/syntax/parser.cr]: Pedantic: That line could be left as is :)

Comment (Maroo-b) [reply src/compiler/crystal/syntax/parser.cr]: @Sija Done :)

**Pull request merged**